exec/cnex/number.c

```diff
@@ -77,79 +91,89 @@ void number_toString(Number x, char *buf, size_t len)
             // of the significant mantissa, adding zeros as necessary.
-            memcpy(r, mantissa, last_significant_digit+1);
+            memcpy(buf, mantissa, last_significant_digit+1);
             for (int i = last_significant_digit+1; i < ((last_significant_digit+1) + exponent); i++, slen++) {
-                r[i] = '0';
+                buf[i] = '0';
             }
         } else if (exponent < 0 && -exponent < slen) {
             // Our exponent is negative, and is less than the length of the mantissa, so we're
             // going to insert the decimal into slen+exponent location.
-            memcpy(r, mantissa, slen+exponent);
-            r[slen+exponent] = '.';
-            memcpy(&r[slen+exponent+1], &mantissa[slen+exponent], slen - (slen+exponent));
+            memcpy(buf, mantissa, slen+exponent);
+            buf[slen+exponent] = '.';
+            memcpy(&buf[slen+exponent+1], &mantissa[slen+exponent], slen - (slen+exponent));
         } else if (exponent < 0 && -exponent == slen) {
             // Our exponent is negative, and is the length of the mantissa, so we're going to start
             // the number with "0.", and then add the rest of the mantissa on the right side of the decimal.
-            memcpy(r, "0.", 2);
-            memcpy(&r[2], mantissa, slen+=2);
-        } else if (exponent < 0 && slen - exponent <= PRECISION + 1) {
+            memcpy(buf, "0.", 2);
+            memcpy(&buf[2], mantissa, slen+=2);
+        } else if (exponent < 0 && (format == nfFull || slen - exponent <= PRECISION + 1)) {
+            if ((size_t)((-exponent-slen) + 2 + slen) > len) {
+                // The caller didn't supply us with enough buffer space to hold the resulting number.
+                return (-exponent-slen) + 2 + slen + 1; // Plus 1.  Don't every forget the +1!  We need that NULL char!
+            }
             // Our exponent is negative, and the mantissa length, minus the exponent less or equal to our max PRECISION
             // then we need to insert zero's to the right of the decimal.
             int count = 2;
-            // Start out by setting [r]esult to "0."
-            memcpy(r, "0.", count);
+            // Start out by setting buf to "0."
+            memcpy(buf, "0.", count);
             // Now, we copy the necessary number of zeros after the decimal point...
             for (int i = 0; i < -exponent-slen;i++, count++) {
-                r[count] = '0';
+                buf[count] = '0';
             }
-            // Then copy the remainder of the mantissa to our [r]esult.
-            memcpy(&r[count], mantissa, slen);
-            // Ensure that [r]esult is null terminated.  I'm not 100% convinced this is necessary.
+            // Then copy the remainder of the mantissa to the supplied buf
+            memcpy(&buf[count], mantissa, slen);
+            // Ensure that buf is null terminated.  I'm not 100% convinced this is necessary.
```

Comment:
Why wouldn't it be necessary to terminate the string? Seems necessary to me.

Reply:
Removed incorrect comment.